crates/phactory/src/system/mod.rs

```diff
@@ -894,127 +1021,26 @@ impl<Platform: pal::Platform> System<Platform> {
-                    (BALANCES => balances::Balances::new()),
-                    (ASSETS => assets::Assets::new()),
-                    (BTC_LOTTERY => btc_lottery::BtcLottery::new(Some(contract_key.0.to_raw_vec()))),
-                    (GEOLOCATION => geolocation::Geolocation::new()),
-                    (GUESS_NUMBER => guess_number::GuessNumber::new()),
-                    (BTC_PRICE_BOT => btc_price_bot::BtcPriceBot::new())
-                };
-
-                self.contract_clusters
-                    .get_cluster_or_default_mut(&cluster_id, &contract_key.0)
-                    .add_contract(contract_id);
-
-                let message = WorkerContractReport::ContractInstantiated {
-                    id: contract_id,
-                    cluster_id,
-                    deployer,
-                    pubkey: EcdhPublicKey(ecdh_pubkey),
-                };
-                info!("Native contract instantiate status: {:?}", message);
-                self.egress.push_message(&message);
-            }
-            CodeIndex::WasmCode(code_hash) => {
-                let code = chain_state::read_contract_code(block.storage, code_hash);
-                if code.is_none() {
-                    error!(
-                        "Code 0x{} not found for contract 0x{}",
-                        hex::encode(code_hash),
-                        hex::encode(contract_key.public())
-                    );
-                    return Err(TransactionError::CodeNotFound.into());
-                }
-
-                let code = code.expect("checked; qed.");
-                let deployer = contract_info.deployer;
-                let effects = self
-                    .contract_clusters
-                    .instantiate_contract(
-                        cluster_id,
-                        deployer.clone(),
-                        code,
-                        contract_info.instantiate_data,
-                        contract_info.salt,
-                        &contract_key,
-                        block.block_number,
-                        block.now_ms,
-                    )
-                    .with_context(|| format!("Contract deployer: {:?}", deployer))?;
-
-                let cluster = self
-                    .contract_clusters
-                    .get_cluster_mut(&cluster_id)
-                    .expect("Cluster must exist after instantiate");
-                apply_pink_side_effects(
-                    effects,
-                    cluster_id,
-                    &mut self.contracts,
-                    cluster,
-                    block,
-                    &self.egress,
-                );
-            }
-        }
+        // TODO(shelven): forget cluster key after expiration time
+        let cluster_key = sr25519::Pair::restore_from_secret_key(&event.secret_key);
+        self.contract_clusters
+            .get_cluster_or_default_mut(&event.cluster, &cluster_key);
```

Comment:
Be careful the key would not be updated if the cluster already exsists.

Reply:
Added an extra check for duplicated deployment.